### ISSUE TITLE
New package to use Earth Engine in R (rgee)

### ISSUE THREAD

**Issue by csaybar** [r-spatial/discuss#33]
Hi everyone!

In the last months, we work in a new package called [rgee](https://csaybar.github.io/rgee). This package hopes to serve as a bridge between google Google Earth Engine (GEE) and the R spatial ecosystem (we develop `rgee` based on how TensorFlow works within in R). This is our first R package, we still need to work harder to enhance the documentation, so before going further we really would appreciate your sincere comments on:

- Is this effort worth it?
-  If so then do you think it makes sense for this proposed package to live in r-spatial?

Pinging @edzer @tim-salabim 

Thanks for any feedback on this!
Cesar

**Comment by edzer** [comment]
Thanks! I tried to start it, but immediately failed; the documentation doesn't mention how to install the `ee` python module needed, and I tried both
```
pip install ee
pip3 install ee
```
but still see
```r
> library(rgee)
> ee$Initialize()
Error: Python module ee was not found.

Detected Python configuration:

python:         /usr/bin/python3
libpython:      /usr/lib/python3.6/config-3.6m-x86_64-linux-gnu/libpython3.6.so
pythonhome:     //usr://usr
version:        3.6.9 (default, Nov  7 2019, 10:44:02)  [GCC 8.3.0]
numpy:          /home/edzer/.local/lib/python3.6/site-packages/numpy
numpy_version:  1.16.4
ee:             /home/edzer/.local/lib/python3.6/site-packages/ee

python versions found: 
 /home/edzer/.virtualenvs/r-reticulate/bin/python
 /usr/bin/python3
 /usr/bin/python
```
some help here, for R users, seems useful.

**Comment by tim-salabim** [comment]
I can't comment much, as I'm not doing much spatial work anymore these days. It sure sounds interesting!

**Comment by pat-s** [comment]
I don't have a strong word here but I see no reason why it should not live in this org :) (don't have time to test it / use case atm but I assume multiple ppl do)

**Comment by BlasBenito** [comment]
As someone doing lots of spatial work these days, I would say that the effort is worthy indeed. For the busy ones, having to learn the GEE language isn't as cost-effective as having an R access to the API would be. Also, I believe this indeed belongs within R-spatial. Thank you very much for giving us this package, I appreciate it a lot!

**Comment by csaybar** [comment]
thanks everyone for your comments!

@edzer to install all the rgee Python dependencies use:

```
library(rgee)
virtualenv_create("rgee", python = "python3.7")
use_virtualenv("rgee")
ee_install_rgee_python_packages()
```
You can also install the EE Python API manually:

```
pip3 install earthengine-api
```
On the other hand, we have created a docker-image, with reproducible examples [here](https://github.com/csaybar/rgee-docker). There is already a preloaded user, so it will not be a problem not to have an active account on Google Earth Engine.

```
docker run -d -p 8787:8787 -e USER=rgee -e PASSWORD=rgee --name rgee-dev csaybar/rgee

```

**Comment by jimoreira** [comment]
It could be great to be able to use this tool inside r. Nowadays for the r spatial users it take some effort to go with python gee API, or even javasript web coding, and then get results inside r to raster, star or sf objects. Other nice scenario is to use your postgresql database, since gee way of using files its only through google drive. Thanks to the one's involved, it would be mucho appreciated!!

**Comment by edzer** [comment]
@csaybar I'm getting
```r
> library(rgee)
> virtualenv_create("rgee", python = "python3.7")
Error in virtualenv_create("rgee", python = "python3.7") : 
  could not find function "virtualenv_create"
> use_virtualenv("rgee")
Error in use_virtualenv("rgee") : 
  could not find function "use_virtualenv"
> ee_install_rgee_python_packages()
Error: could not find a Python environment for /usr/bin/python3
```

**Comment by edzer** [comment]
When running your docker image, and after installing `rgee` in it (!), I still see
```
> library(rgee)
> ee$Initialize()
Error: Python module ee was not found.

Detected Python configuration:

python:         /usr/bin/python3
libpython:      /usr/lib/python3.5/config-3.5m-x86_64-linux-gnu/libpython3.5.so
pythonhome:     /usr:/usr
version:        3.5.3 (default, Sep 27 2018, 17:25:39)  [GCC 6.3.0 20170516]
numpy:          /usr/local/lib/python3.5/dist-packages/numpy
numpy_version:  1.17.2
ee:             /usr/local/lib/python3.5/dist-packages/ee

NOTE: Python version was forced by RETICULATE_PYTHON
```

**Comment by csaybar** [comment]
Hi @edzer,

I apologize, the correct code would be:

``` r
library(rgee)
library(reticulate)
virtualenv_create("rgee", python = "python3.7")
use_virtualenv("rgee")
ee_install_rgee_python_packages()
```
With respect to the docker image:

``` r
library(rgee)
ee_reattach() #reattach ee as a reserve word
ee_Initialize()
```
You could use ee$Initialize() also, however you would need to authenticate manually first running in the console:

``` bash
!earthengine authenticate
```

I've just made the changes, thanks a lot!

**Comment by edzer** [comment]
I still see in the docker image
```
> library(rgee)
> ee_reattach() #reattach ee as a reserve word
> ee_Initialize()
Error: Python module ee was not found.

Detected Python configuration:

python:         /usr/bin/python3
libpython:      /usr/lib/python3.5/config-3.5m-x86_64-linux-gnu/libpython3.5.so
pythonhome:     /usr:/usr
version:        3.5.3 (default, Sep 27 2018, 17:25:39)  [GCC 6.3.0 20170516]
numpy:          /usr/local/lib/python3.5/dist-packages/numpy
numpy_version:  1.17.2
ee:             /usr/local/lib/python3.5/dist-packages/ee

NOTE: Python version was forced by RETICULATE_PYTHON
```
and on the command line
```
rgee@c737af85f56b:~$ !earthengine authenticate
bash: !earthengine: event not found
```

**Comment by kanedan29** [comment]
This effort is definitely worth it! I'd echo @BlasBenito and add that GEE is limited in the analytical functions it offers. Being able to access GEE's super useful API and then apply the R spatial ecosystem will be awesome.

**Comment by csaybar** [comment]
@edzer I see the problem. Apparently, there have been some errors when building the docker, and building/uploading it will take me many hours (due to my internet connection, but I started to do it!). However, the error that you mentioned could be solved running:

``` r
install.packages('rgee')
library(rgee)
system('pip3 install --upgrade pyasn1-modules')
rstudioapi::restartSession()
```
The demo folder is available [here](https://github.com/csaybar/rgee-docker/tree/master/demo). Please let me know if the problem continues, thank you for the feedback!

**Comment by csaybar** [comment]
@kanedan29  Yes, that is the idea!. For instance, you could go from Earth Engine to R using [ee_as_sf](https://csaybar.github.io/rgee/reference/ee_as_sf.html) and vice versa ([sf_as_ee](https://csaybar.github.io/rgee/reference/sf_as_ee.html))

**Comment by ctlamb** [comment]
I would find this extremely useful for my work. Thanks

**Comment by tim-salabim** [comment]
It sure seems to be popular with the people of twitter... My feed is going nuts today

https://twitter.com/TimSalabim3/status/1219170244138295296

**Comment by edzer** [comment]
@csaybar that seems to get me beyond the authentication step, but then if I try the nightlight example I get
```
> collection = ee$
+     ImageCollection('NOAA/DMSP-OLS/NIGHTTIME_LIGHTS')$
+     select('stable_lights')$
+     map(createTimeBand)
 Error in py_call_impl(callable, dots$args, dots$keywords) : 
  EEException: Server returned HTTP code: 404 
```
@kanedan29 please try this out if this has potential for you. I'm pretty sure that the R things you can concerns data you download from GEE, you won't be able to push R functions to be evaluated inside GEE.

**Comment by csaybar** [comment]
@edzer It is really strange, did you try the last docker image that I upload?
```
docker pull csaybar/rgee:latest
```
It should work fine now.

![ep](https://user-images.githubusercontent.com/16768318/72761095-e9b94880-3ba8-11ea-99a5-9a061ca3b3e9.png)

**Comment by pat-s** [comment]
I'd suggest to move such specific debugging discussions to a separate issue?

**Comment by edzer** [comment]
Patrick, I appreciate your suggestions, but I find that before we move this package to the r-spatial org, at least one person (beyond the author) should report successful installation & running examples. That is what I am trying, and as long as I am not succeeding and I don't hear anyone else succeed, I am keeping the discussion here. Beyond comments, your help would be appreciated too.

**Comment by kanedan29** [comment]
@csaybar @edzer, I was able to get the docker image to work and ran the night-time lights example (see attached image). Was struggling with setup on my local machine though - seems like just some python configuration issues that I'll have to work through. Would be happy to share insights from that process here or in another thread. 

<img width="1552" alt="image" src="https://user-images.githubusercontent.com/9805273/72826471-e4aad680-3c46-11ea-9f1a-350ce6eb7ad6.png">

**Comment by csaybar** [comment]
@edzer we sorry difficulties, we will continue working on the installation and authentication module to make it more intuitive. The documentation is quite unclear also, we hope to have everything ready in version 0.2.0.  I do not know if it is appropriate to move now or after the release of version 0.2.0 rgee to r-spatial but in both cases, we have no problem!.

**Thank you very much, everyone, for your comments!** they help and motivate us to continue working on this project.

**Comment by edzer** [comment]
Looks great - I also managed to run examples in the new docker container, but it no longer asks for my google credentials - does it use a pre-specified user's credentials? I still want to be able to run this outside the container, will try more later.

**Comment by pat-s** [comment]
> Patrick, I appreciate your suggestions, but I find that before we move this package to the r-spatial org, at least one person (beyond the author) should report successful installation & running examples. That is what I am trying, and as long as I am not succeeding and I don't hear anyone else succeed, I am keeping the discussion here. 

Fair enough; for me install problems which are followed by longer discussion(s) (let's say more than 2 comments) about a particular problem have the potential to make an issue somewhat messy when the initial topic had a different intention.

Users arriving here will have to read a lot of specific comments (including mine) which might not apply anymore in some days even and do not relate to discussing the strategic future of the package.

(Hiding those by marking them as "off-topic" is one potential way to clean up)

Not sure how my comment was perceived; I had the feeling to explain it 😄 

> Beyond comments, your help would be appreciated too.

Thanks, but sadly I have no time for this atm (and never had a use case for EE so far). Strictly limiting myself atm, probably I should even turn off GH notifications at all for some time :(

---

Great to see such an effort in the r-spatial com, highly appreciated!

**Comment by csaybar** [comment]
> Looks great - I also managed to run examples in the new docker container, but it no longer asks for my google credentials - does it use a pre-specified user's credentials? I still want to be able to run this outside the container, will try more later.
@edzer 
Yes, you just need to Initialize once.  `rgee`  offers also multi-user support:
``` r
ee_Initialize(user_gmail = 'csaybar@gmail.com')  # just change it by your email
```
The credentials are stored in:
``` r
path.expand("~/.config/earthengine/")
```
If `reticulate` is the problem maybe this provisional post [help](https://csaybar.github.io/rgee/articles/faq.html)

**Comment by ctlamb** [comment]
Hello, really excited about this package! I am having similar issues to @edzer. I downloaded Python, and followed the trouble shooting steps outlined in this thread. I have the Docker running, but I mist admit I am not fluent in Docker approaches. Would certainly prefer to run outside of Docker image.

`ee_Initialize()
Error: Python module ee was not found.

Detected Python configuration:

python:         /usr/bin/python
libpython:      /System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/config/libpython2.7.dylib
pythonhome:     /System/Library/Frameworks/Python.framework/Versions/2.7:/System/Library/Frameworks/Python.framework/Versions/2.7
version:        2.7.16 (default, Nov  9 2019, 05:55:08)  [GCC 4.2.1 Compatible Apple LLVM 11.0.0 (clang-1100.0.32.4) (-macos10.15-objc-s
numpy:          /System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/numpy
numpy_version:  1.8.0

python versions found: 
 /usr/bin/python
 /usr/bin/python3
 /usr/local/bin/python3`

**Comment by csaybar** [comment]
Hi @ctlamb, `rgee` only is compatible with Python version 3.5 >=. (you are using Python 2.7).
It is easy to fix since you are using a Unix system. Firstly, try to update your installation of `rgee` using:
``` r
remotes::install_github("csaybar/rgee")
```
And follow this tutorial:

<img src='https://user-images.githubusercontent.com/16768318/74463774-56e69380-4e60-11ea-98c9-10c1662a5bf4.png' width='80%'>

**We will happy to help you with additional issues, but please open an issue [here](https://github.com/csaybar/rgee/issues)**

**Comment by csaybar** [comment]
Hi everyone, I am happy to announce the version 0.2.0 of rgee. Now, the [installation module](https://github.com/csaybar/rgee#requirements) is more intuitive and user-friendly. In addition, several bugs were fixed. We consider that it would be time to transfer the ownership [rgee](https://github.com/csaybar/rgee) to r-spatial.org. 

In case of any problem, do not dude in open an [issue](https://github.com/csaybar/rgee/issues).

**Comment by tim-salabim** [comment]
Thanks, I won't have time to try this. But it just occured to me if it would make sense to streamline the python environment related code bits with those from [RQGIS3](https://github.com/r-spatial/RQGIS3)? Not that I have a good idea about all this, but maybe @jannes-m and/or @pat-s can weigh in here?

**Comment by GMinRivrs** [comment]
Hello, i have been using this package and it works great so far for me, however it fails when i try to process images in parallel, and every function fails saying 'attempt to apply non-function' whenever i try to map images (ex ee$image$clip() for example). i wonder if there is something i can do to fix this. [the package registers correctly on each core, but the functions don't work]

**Comment by GMinRivrs** [comment]
getpercNDVI=function(turfForUser,NDVInonIrri10,NDVInonIrri90){
#class(NDVInonIrri10) 
#[1] "sf"         "data.frame"
#class(NDVInonIrri10)
#[1] "ee.image.Image"                   "ee.element.Element"               "ee.computedobject.ComputedObject" "ee.encodable.Encodable"          
#[5] "python.builtin.object" 
#

  if(nrow(turfForUser)<80){ #if i have more than 80 features, run the analysis
    MSK=sf_as_ee(turfForUser)
   
   if(as.numeric(sum(st_area(turfForUser)))<1){sc=0.1}else{sc=10} #if the area of the feature is less than 1m, change the scale
      
    ee_min = NDVInonIrri10$reduceRegion(
      reducer = ee$Reducer$median(),
      geometry = MSK,
      scale = sc,
      maxPixels = 1e9
    )$getInfo()$nd
    
    ee_max= NDVInonIrri90$reduceRegion(
      reducer = ee$Reducer$median(),
      geometry = MSK,
      scale = sc,
      maxPixels = 1e9
    )$getInfo()$nd
	
	PPNDVI=c(ee_max,ee_min)#
  return(PPNDVI)
}

note that this exact same function works if i apply locally, but if i apply it on a parallel way, it doesn't

**Comment by GMinRivrs** [comment]
@edzer i think the problem is reticulate, not Rgee itself. going more in details with the errors, it appears that the actual problem is that reticulate creates objects that can not be exported to other processes [''Error: Detected a non-exportable reference (‘externalptr’) in one of the globals (‘NDVInonIrri10’ of class ‘ee.image.Image’)'' ], so i need to find a way to create an instance of reticulate in each cluster, to make rgee work...

**Comment by edzer** [comment]
It sounded like you wanted to distribute your process on the client-side, where GEE distributes it on the server side. If this is the case, I'm not sure whether this is meant to work out, from a GEE perspective.

**Comment by csaybar** [comment]
You will not obtain an improvement in time by parallelizing your process at the client-side. However, if you are working in collaboration with users with active GEE accounts a code like this could help you to obtain results three-times faster.

```r
library(foreach)
library(rgee)
google_account <- c("csaybar", "ryali93", "lbautista")
foreach(account = google_account, .combine = "c") %dopar% {
  ee_Initialize(account)
  ic_results <- temporal_deforestation(split = account)
  ee_imagecollection_to_local(ic_results)
} -> results
```
Please if you continue having problems open an issue [here](https://github.com/csaybar/rgee/issues)

**Comment by csaybar** [comment]
Hi everyone, we hope you are well!

We kept continuously working in [rgee](https://github.com/csaybar/rgee) all this time and now we have a stable version.  We think is time to transfer rgee to r-spatial.org.  Since our plan is to release the package on CRAN in a few weeks. 

If the transfer is possible. How should I proceed?

@edzer @tim-salabim

**Comment by tim-salabim** [comment]
Sorry for the silence. I have no problems with moving rgee here, though I haven't tried it myself so far. 

@edzer?

**Comment by edzer** [comment]
Yes, I'm good to go. I think the way to do this is "settings" -> "transfer ownership".

**Comment by csaybar** [comment]
Hi, @edzer  @tim-salabim thanks for the answer. It seems I need an invitation to transfer rgee to r-spatial. Would you also settings access for rgee to @ryali93?. We will administrate the package. 
![image](https://user-images.githubusercontent.com/16768318/81991401-c9554400-9606-11ea-9e64-dc08e6f78582.png)

**Comment by tim-salabim** [comment]
@csaybar I just invited you and @ryali93

**Comment by csaybar** [comment]
@tim-salabim @edzer
I made the transfer thanks, just a last petition: would some of you give us push/pull access and secret settings to rgee?

**Comment by edzer** [comment]
You should have commit/push access already.

**Comment by csaybar** [comment]
Thanks @edzer everything works perfectly!. Just to finish, I have no access to change the URL of rgee here. 

![image](https://user-images.githubusercontent.com/16768318/82106196-2293a500-96e5-11ea-8597-8c839f2ace49.png)

Could I have access to change the URL from csaybar to r-spatial, or by defect someone could change it?

**Comment by edzer** [comment]
I now did that. But otherwise, you have full access to the repo?

**Comment by csaybar** [comment]
I do not have access to the settings section. but everything else I have full access

**Comment by tim-salabim** [comment]
@csaybar @ryali93 I've just made you admins of the rgee repo.

**Comment by csaybar** [comment]
Thanks so much @tim-salabim, we have access to access to settings now!

**Comment by Yinshao-lzu** [comment]
hello, I just started learning the Rgee package, but an error is trouble me. Could you tell me the solution? Thanks.
![QQ截图20200828114819](https://user-images.githubusercontent.com/56902640/91519173-763c9e00-e924-11ea-8e33-b47e10e7cc7a.png)

**Comment by csaybar** [comment]
Hi @Yinshao-lzu thanks for your interest in rgee, please describe the issue here https://github.com/r-spatial/rgee/issues.
